[RFC] Test: fix functional/ex_cmds/recover_spec.lua

mhinz opened the pull request:
Commit 1: add `helpers.rmdir()`
Commit 2: fix bug in `test/functional/ex_cmds/recover_spec.lua` that lead to a stray directory

mhinz committed:
Test: add new helper function: rmdir()

- lfs.rmdir() only removes empty directories

- os.remove() supercedes lfs.rmdir(); removes files and empty directories

- helpers.rmdir() first removes all files within a directory, then the
  directory itself

tarruda commented:
:+1:

justinmk commented on test/functional/helpers.lua:
```diff
@@ -290,6 +290,22 @@ local function expect(contents)
+    if file == '.' or file == '..' then
+      goto continue
+    end
+    if not os.remove(path..'/'..file) then
```
is there a way to print or return the error?

mhinz replied:
Yup, both would be possible.

Caller handles problems:
```lua
ret, err = func()
if not ret then
  return nil, err
end
```
Callee handles problems:
```lua
ret, err = func()
if not ret then
  print('func(): '..err)
  return nil
end
```
Or a combination of course. I prefer the callee solution, so the tests don't get cluttered with conditionals.

justinmk replied:
Either works for me, but it should probably throw an error so that the test fails.

mhinz replied:
Oh, so `error()` works as a `print()`, but counts as error. I didn't mention it previously because I thought the whole test would fail immediately, which is not what we want, because we would like to  execute `after_each()` etc. But.. that all seems to work just fine.

I'll change that.

mhinz committed:
Test: fix functional/ex_cmds/recover_spec.lua

os.remove() wasn't removing the temporary swap directory which leads to
problems when the test is run a second time.

That's also the reason why the CI never caught this.

os.remove() got replaced by helpers.rmdir().

justinmk commented:
LGMT, thanks!

Pyrohh commented:
How is `helpers.rmdir` different from [`lfs.rmdir`](https://keplerproject.github.io/luafilesystem/manual.html)?

fmoralesc commented:
I would bet `lfs.rmdir` fails if the dir is not empty.

justinmk commented:
Yep, just like `rmdir` in the shell. Even .NET doesn't have a "nuke directory" function, it seems like framework authors are generally wary of providing such a thing.

fmoralesc commented:
I think it's because POSIX defines rmdir like that.

justinmk commented:
In fact this reminds me that we should probably add a sanity check in `helpers.rmdir()` so that contributors don't accidentally delete something outside of the neovim source tree.
